File: Haskell/04-straight-flush.hs
import Data.Char(digitToInt)
import Data.List (nub, sort)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as M

data Suit = Spades | Hearts | Diamonds | Clubs deriving Eq
data Card = Card {suit :: Suit, rank :: Int}


rankToInt :: Map Char Int
rankToInt = M.fromList [('A', 14), ('K', 13), ('Q', 12), ('J', 11), ('T', 10)]


readCards :: String -> [Card]
readCards line = map (\[r, s] -> Card (charToSuit s) (rank r)) $ words line
    where
        charToSuit :: Char -> Suit
        charToSuit c
            | c == 'S' = Spades
            | c == 'H' = Hearts
            | c == 'D' = Diamonds
            | otherwise = Clubs
        rank :: Char -> Int
        rank r = case M.lookup r rankToInt of Nothing -> digitToInt r
                                              Just n -> n


isStraightFlush :: [Card] -> Bool
isStraightFlush cards =
    let suits = nub $ map (\(Card suit _) -> suit) cards
        checkRankDifferences :: Bool
        checkRankDifferences =
            let ranks = sort $ map (\(Card _ rank) -> rank) cards
                differences = map (\(b, a) -> b - a) $ zip (tail ranks) ranks
            in differences == [1, 1, 1, 1] || differences == [1, 1, 1, 9]
    in if length suits /=1 then False else checkRankDifferences


main :: IO()
main = do
    cards <- fmap readCards getLine
    let result = isStraightFlush cards
    putStrLn $ if result then "YES" else "NO"

File: Python/04_straight-flush.py
import sys

from collections import namedtuple


Card = namedtuple('Card', ['suit', 'rank'])


def read_cards(line):
    cards = []
    for s in line.split():
        suit = s[-1]
        rank = convert_rank(s[:-1])
        cards.append(Card(suit, rank))
    return cards


def convert_rank(rank):
    if rank == 'A':
        return 14
    if rank == 'K':
        return 13
    if rank == 'Q':
        return 12
    if rank == 'J':
        return 11
    if rank == 'T':
        return 10
    return int(rank)


def is_straight_flush(cards):
    suits = {card.suit for card in cards}
    if len(suits) != 1:
        return False

    ranks = sorted([card.rank for card in cards])
    differences = [a - b for a, b in zip(ranks[1:], ranks)]
    return (differences == [1, 1, 1, 1] or differences == [1, 1, 1, 9])


def main():
    data = sys.stdin.read().splitlines()
    cards = read_cards(data[0])
    result = is_straight_flush(cards)
    print('YES' if result else 'NO')


if __name__ == '__main__':
    main()
